### PR TITLE
Add initial Elixir suite

### PR DESCRIPTION
**Pull request by chewbranca**
## Overview

This is a PR for the current state of the Elixir test suite. This is a WIP, but it's functional enough for folks to run with, and there's a desire for new tests to be written in it. This isn't yet connected to `make check`, so I'm opening up this PR with the intention of merging as is to allow more folks to start diving in for porting the existing tests and creating new tests. We can streamline things a bit more and then get it properly connected in.

Current status of running the `./test/elixir/run` command:

```
112 tests, 4 failures, 5 skipped 
```

## Testing recommendations

```
# install elixir and mix
cd test/elixir
./run
```

## Checklist

- [ ] Code is written and works correctly;
- [ ] Changes are covered by tests;
- [ ] Documentation reflects the changes;

### CODE REVIEW

**Comment by wohali** [comment]
After `apt-get install elixir`, I end up with:

```
$ ./run
Could not find Hex, which is needed to build dependency :httpotion
Shall I install Hex? (if running non-interactively, use: "mix local.hex --force") [Yn]
```

I added this command to the top of ./run and got:

```
** (Mix) You're trying to run :foo on Elixir v1.3.3 but it has declared in its mix.exs file it supports only Elixir ~> 1.5
```

meaning that the Debian version of elixir isn't compatible with the newer packaged versions of Erlang from erlang-solutions.

I'll have to come back to this, I'm out of time for Couch until Monday.

**Comment by garrensmith** [comment]
The tests are all green on this. I would be keen to merge this. Is there anything other than maybe cleaning up the commits stopping us from merging this?

**Comment by wohali** [comment]
@garrensmith let me cut the 2.2.x branch first - should be within 24-48h - then go for it

**Comment by iilyak** [comment]
> wohali: meaning that the Debian version of elixir isn't compatible with the newer packaged versions of Erlang from erlang-solutions.

@wohali an attempt to solve this issue is in https://github.com/apache/couchdb/pull/1466. Unfortunately windows side of things most likely doesn't work and need some love. I might get to it, but cannot promise.

File: Makefile [hunk 1]
```diff
@@ -137,6 +137,11 @@ soak-eunit: couch
 	@$(REBAR) setup_eunit 2> /dev/null
 	while [ $$? -eq 0 ] ; do $(REBAR) -r eunit $(EUNIT_OPTS) ; done
 
+.PHONY: elixir
+elixir:
+	@rm -rf dev/lib
+	@dev/run -a adm:pass --no-eval test/elixir/run
```
Comment (wohali) [inline Makefile]: So it is impossible from an Elixir test to restart CouchDB? That is going to limit some of the kinds of tests we do in the JavaScript suite.

Why was this done, out of curiosity?

File: src/chttpd/src/chttpd_misc.erl [hunk 2]
```diff
@@ -240,7 +240,9 @@ cancel_replication(PostBody, Ctx) ->
             {error, badrpc};
         Else ->
             % Unclear what to do here -- pick the first error?
-            hd(Else)
+            % Except try ignoring any {error, not_found} responses
+            % because we'll always get two of those
+            hd(Else -- [{error, not_found}])
```
Comment (wohali) [inline src/chttpd/src/chttpd_misc.erl]: Unrelated change - can we move this out of this PR please?

File: Makefile (hunk 1 above)
Comment (iilyak) [reply Makefile]: I think we should be able to fix it. There are few options available. It is a matter of choosing the best one:
  - run script which reads commands from stdin (as it is done for JS tests)
  - similar thing in elixir which restarts beam and haproxy
  - simulation of a restart using `application:stop(couch)`/`application:start(couch)`
  - http endpoint `_admin/restart` or `_test/restart` (this would require some supervisor to deal with restart)

Elixir test suite is an integration test. When we run integration tests there is a need to do certain operations to facilitate black box testing. Namely SUT restart and failure injection. It would be nice to be able to reuse this test suite to test the installation (before we allow customers traffic) using integration test suite just to make sure everything is in order. Having said that I think that the best option would be to implement the following:
- add `--keep-runing` or `--with-restart` or `--supervise` option to `dev/run` script
- implement http endpoint (TBD) to control the beam from integration test suites.
  - we need to be able to disable this endpoint in production via either:
      - config entry
      - absence of environment variable
      - some other method which would make it hard to enable it by accident. 

This way we would have a similar restart behavior as in production. Since In production we run CouchDB from standard distribution specific supervisor which deals with restart.

Comment (wohali) [reply Makefile]: @iilyak I agree, I'd prefer not to use `application:stop/start` because that's not how CouchDB actually restarts in production most of the time. 

There used to be a restart endpoint, it was dropped with 2.x. The hack in `dev/run` is what @davisp put in to deal with its removal.  If you do put it back in, it should live under `/_node/<nodename>/_restart`.

Comment (iilyak) [reply Makefile]: `POST /_node/<nodename>/_restart` is implemented here https://github.com/apache/couchdb/pull/1543

**Comment by janl** [comment]
Getting errors when running this, is there anything obvious I’m missing?

I don’t think it is related, but I’ve had to do a PATH hack to allow CouchDB to be built with Erlang 20 and Elixir to ru under Erlang 21 because Homebrew.

```
> elixir --erl -i
jan@rose (12:28:30) ~/Work/couchdb (elixir-suite) 
> make elixir
[ * ] Setup environment ... ok
[ * ] Ensure CouchDB is built ... ok
[ * ] Prepare configuration files ... ok
[ * ] Start node node1 ... ok
[ * ] Start node node2 ... ok
[ * ] Start node node3 ... ok
[ * ] Check node at http://127.0.0.1:15984/ ... ok
[ * ] Check node at http://127.0.0.1:25984/ ... ok
[ * ] Check node at http://127.0.0.1:35984/ ... ok
[ * ] Running cluster setup ... ok
[ * ] Exec command test/elixir/run ... Resolving Hex dependencies...
Dependency resolution completed:
  httpotion 3.0.3
  ibrowse 4.4.0
  jiffy 0.14.11
All dependencies up to date
===> Compiling jiffy
Compiling c_src/decoder.c
Compiling c_src/encoder.c
Compiling c_src/jiffy.c
Compiling c_src/utf8.c
Compiling c_src/util.c
Compiling c_src/doubles.cc
Compiling c_src/double-conversion/bignum-dtoa.cc
Compiling c_src/double-conversion/bignum.cc
Compiling c_src/double-conversion/cached-powers.cc
Compiling c_src/double-conversion/diy-fp.cc
Compiling c_src/double-conversion/double-conversion.cc
Compiling c_src/double-conversion/fast-dtoa.cc
Compiling c_src/double-conversion/fixed-dtoa.cc
Compiling c_src/double-conversion/strtod.cc
===> Compiling ibrowse
==> httpotion
Compiling 1 file (.ex)
Generated httpotion app
==> foo
Compiling 1 file (.ex)
Generated foo app
warning: found quoted keyword "_id" but the quotes are not required. Note that keywords are always atoms, even when quoted, and quotes should only be used to introduce keywords with foreign characters in them
  test/test_helper.exs:178

warning: found quoted keyword "bar" but the quotes are not required. Note that keywords are always atoms, even when quoted, and quotes should only be used to introduce keywords with foreign characters in them
  test/test_helper.exs:179

Excluding tags: [pending: true]


== Compilation error in file test/compact_test.exs ==
** (CompileError) test/compact_test.exs:83: defp rev/2 already defined as def
    test/compact_test.exs:83: (module)
    (stdlib) erl_eval.erl:677: :erl_eval.do_apply/6
    (elixir) lib/code.ex:767: Code.require_file/2
    (elixir) lib/kernel/parallel_compiler.ex:209: anonymous fn/4 in Kernel.ParallelCompiler.spawn_workers/6
make: *** [elixir] Error 1
```

File: src/chttpd/src/chttpd_misc.erl (hunk 2 above)
Comment (janl) [reply src/chttpd/src/chttpd_misc.erl]: Done in #174677fc8

Comment (janl) [reply src/chttpd/src/chttpd_misc.erl]: and https://github.com/apache/couchdb/pull/1614

**Comment by jaydoane** [comment]
@janl that compiler error is my bad, sorry! Will fix today.

**Comment by wohali** [comment]
@chewbranca @jaydoane I'm ready to merge this once tests have passed. Can you please resolve the conflict in `chttpd_misc.erl`?

**Comment by wohali** [comment]
Never mind, I'll fix it myself...
